// @assistant-ui/react/dist/context/providers/MessageProvider.js
"use client";

// src/context/providers/MessageProvider.tsx
import { useMemo } from "react";
import {
  AssistantProvider,
  createAssistantApiField
} from "../react/AssistantApiContext.js";
import {
  resource,
  tapInlineResource,
  tapMemo,
  tapState
} from "@assistant-ui/tap";
import { useResource } from "@assistant-ui/tap/react";
import { asStore, tapApi } from "../../utils/tap-store/index.js";
import { tapLookupResources } from "../../client/util-hooks/tapLookupResources.js";
import { jsx } from "react/jsx-runtime";
var NoOpComposerClient = resource(({ type }) => {
  const state = tapMemo(() => {
    return {
      isEditing: false,
      isEmpty: true,
      text: "",
      attachmentAccept: "*",
      attachments: [],
      role: "user",
      runConfig: {},
      canCancel: false,
      type
    };
  }, [type]);
  return tapApi({
    getState: () => state,
    setText: () => {
      throw new Error("Not supported");
    },
    setRole: () => {
      throw new Error("Not supported");
    },
    setRunConfig: () => {
      throw new Error("Not supported");
    },
    addAttachment: () => {
      throw new Error("Not supported");
    },
    clearAttachments: () => {
      throw new Error("Not supported");
    },
    attachment: () => {
      throw new Error("Not supported");
    },
    reset: () => {
      throw new Error("Not supported");
    },
    send: () => {
      throw new Error("Not supported");
    },
    cancel: () => {
      throw new Error("Not supported");
    },
    beginEdit: () => {
      throw new Error("Not supported");
    },
    __internal_getRuntime: () => null
  });
});
var ThreadMessagePartClient = resource(
  ({ part }) => {
    const state = tapMemo(() => {
      return {
        ...part,
        status: { type: "complete" }
      };
    }, [part]);
    return tapApi(
      {
        getState: () => state,
        addToolResult: () => {
          throw new Error("Not supported");
        },
        __internal_getRuntime: () => null
      },
      {
        key: state.type === "tool-call" ? "toolCallId-" + state.toolCallId : void 0
      }
    );
  }
);
var ThreadMessageAttachmentClient = resource(
  ({ attachment }) => {
    return tapApi(
      {
        getState: () => attachment,
        remove: () => {
          throw new Error("Not supported");
        },
        __internal_getRuntime: () => null
      },
      {
        key: attachment.id
      }
    );
  }
);
var ThreadMessageClient = resource(
  ({
    message,
    isLast = true,
    branchNumber = 1,
    branchCount = 1
  }) => {
    const [isCopiedState, setIsCopied] = tapState(false);
    const [isHoveringState, setIsHovering] = tapState(false);
    const parts = tapLookupResources(
      message.content.map(
        (_, idx) => ThreadMessagePartClient({ part: message.content[idx] }, { key: idx })
      )
    );
    const attachments = tapLookupResources(
      message.attachments?.map(
        (_, idx) => ThreadMessageAttachmentClient(
          { attachment: message.attachments[idx] },
          { key: idx }
        )
      ) ?? []
    );
    const composerState = tapInlineResource(
      NoOpComposerClient({ type: "edit" })
    );
    const state = tapMemo(() => {
      return {
        ...message,
        parts: parts.state,
        composer: composerState.state,
        parentId: null,
        isLast,
        branchNumber,
        branchCount,
        speech: void 0,
        submittedFeedback: void 0,
        isCopied: isCopiedState,
        isHovering: isHoveringState
      };
    }, [message, isCopiedState, isHoveringState, isLast]);
    return tapApi({
      getState: () => state,
      composer: composerState.api,
      part: (selector) => {
        if ("index" in selector) {
          return parts.api({ index: selector.index });
        } else {
          return parts.api({ key: "toolCallId-" + selector.toolCallId });
        }
      },
      attachment: (selector) => {
        if ("id" in selector) {
          return attachments.api({ key: selector.id });
        } else {
          return attachments.api(selector);
        }
      },
      reload: () => {
        throw new Error("Not supported in ThreadMessageProvider");
      },
      speak: () => {
        throw new Error("Not supported in ThreadMessageProvider");
      },
      stopSpeaking: () => {
        throw new Error("Not supported in ThreadMessageProvider");
      },
      submitFeedback: () => {
        throw new Error("Not supported in ThreadMessageProvider");
      },
      switchToBranch: () => {
        throw new Error("Not supported in ThreadMessageProvider");
      },
      getCopyText: () => {
        return message.content.map((part) => {
          if ("text" in part && typeof part.text === "string") {
            return part.text;
          }
          return "";
        }).join("\n");
      },
      setIsCopied,
      setIsHovering,
      __internal_getRuntime: () => null
    });
  }
);
var MessageProvider = ({ children, ...props }) => {
  const store = useResource(asStore(ThreadMessageClient(props)));
  const api = useMemo(() => {
    return {
      message: createAssistantApiField({
        source: "root",
        query: {},
        get: () => store.getState().api
      }),
      subscribe: store.subscribe,
      flushSync: store.flushSync
    };
  }, [store]);
  return /* @__PURE__ */ jsx(AssistantProvider, { api, children });
};
export {
  MessageProvider
};
