// assistant-stream/dist/core/index.js
// src/core/index.ts
import {
  createAssistantStream,
  createAssistantStreamResponse,
  createAssistantStreamController
} from "./modules/assistant-stream.js";
import {
  AssistantMessageAccumulator,
  createInitialMessage
} from "./accumulators/assistant-message-accumulator.js";
import { AssistantStream } from "./AssistantStream.js";
import {
  DataStreamDecoder,
  DataStreamEncoder
} from "./serialization/data-stream/DataStream.js";
import { PlainTextDecoder, PlainTextEncoder } from "./serialization/PlainText.js";
import { AssistantMessageStream } from "./accumulators/AssistantMessageStream.js";
export * from "./tool/index.js";
import { createObjectStream } from "./object/createObjectStream.js";
import {
  ObjectStreamResponse,
  fromObjectStreamResponse
} from "./object/ObjectStreamResponse.js";
export {
  AssistantMessageAccumulator,
  AssistantMessageStream,
  AssistantStream,
  DataStreamDecoder,
  DataStreamEncoder,
  ObjectStreamResponse,
  PlainTextDecoder,
  PlainTextEncoder,
  createAssistantStream,
  createAssistantStreamController,
  createAssistantStreamResponse,
  createObjectStream,
  fromObjectStreamResponse,
  createInitialMessage as unstable_createInitialMessage
};
