// @assistant-ui/react/dist/internal.js
// src/internal.ts
import { DefaultThreadComposerRuntimeCore } from "./legacy-runtime/runtime-cores/composer/DefaultThreadComposerRuntimeCore.js";
import { CompositeContextProvider } from "./utils/CompositeContextProvider.js";
import { MessageRepository } from "./legacy-runtime/runtime-cores/utils/MessageRepository.js";
import { BaseAssistantRuntimeCore } from "./legacy-runtime/runtime-cores/core/BaseAssistantRuntimeCore.js";
import { generateId } from "./utils/idUtils.js";
import { AssistantRuntimeImpl } from "./legacy-runtime/runtime/AssistantRuntime.js";
import {
  ThreadRuntimeImpl
} from "./legacy-runtime/runtime/ThreadRuntime.js";
import { fromThreadMessageLike } from "./legacy-runtime/runtime-cores/external-store/ThreadMessageLike.js";
import { getAutoStatus } from "./legacy-runtime/runtime-cores/external-store/auto-status.js";
import { splitLocalRuntimeOptions } from "./legacy-runtime/runtime-cores/local/LocalRuntimeOptions.js";
export * from "./utils/smooth/index.js";
export {
  AssistantRuntimeImpl,
  BaseAssistantRuntimeCore,
  CompositeContextProvider,
  DefaultThreadComposerRuntimeCore,
  MessageRepository,
  ThreadRuntimeImpl,
  fromThreadMessageLike,
  generateId,
  getAutoStatus,
  splitLocalRuntimeOptions
};
